Add `padding_idx` argument to EmbeddingBag

kurtamohler opened the pull request:
This PR adds a `padding_idx` parameter to `nn.EmbeddingBag` and `nn.functional.embedding_bag`. As with `nn.Embedding`'s `padding_idx` argument, if an embedding's index is equal to `padding_idx` it is ignored, so it is not included in the reduction.

This PR does not add support for `padding_idx` for quantized or ONNX `EmbeddingBag` for opset10/11 (opset9 is supported). In these cases, an error is thrown if `padding_idx` is provided.


Fixes #3194

kurtamohler committed:
Add `padding_idx` argument to EmbeddingBag

kurtamohler commented:
There are several more changes I need to make, including updating the C++ interface in `torch/csrc/api/src/nn/` and adding test cases.

kurtamohler commented:
I'm still working on fixing the quantization test failures. My current plan is to add the `padding_idx` argument, but throw an error if the user actually provides it. Full support for quantized `padding_idx` can be added in a follow up PR.

kurtamohler commented on torch/csrc/autograd/variable.h:
```diff
@@ -248,12 +245,25 @@ struct TORCH_API AutogradMeta : public c10::AutogradMetaInterface {
+  // a virtual function in a constructor will call the base override anyway,
+  // but it also causes a clang-tidy error. Calling it directly prevents the
+  // error.
+  void set_requires_grad_base(bool requires_grad, at::TensorImpl* self_impl) {
```
I'm not sure why, but I was getting clang-tidy errors here, even though I wasn't changing any of the surrounding code. I can revert these changes if needed.

peterbell10 commented on torch/onnx/symbolic_opset11.py:
```diff
@@ -893,9 +893,12 @@ def embedding_bag(g,
     if scale_grad_by_freq and sym_help._training_mode:
         return sym_help._onnx_unsupported('embedding_bag with scale_grad_by_freq for training mode')
+    if padding_idx is not None and padding_idx >= 0:
+        raise RuntimeError('embedding_bag with padding_idx')
```
IIRC the onnx fallback operators are generated from the `native_functions.yaml` schema, so is it possible this would just work if you added `padding_idx`to the `g.op` call below?

kurtamohler replied:
For opset9, yes, it does seem to be as simple as adding `padding_idx` to the `g.op("ATen", ..., operator_s="embedding_bag", ...)` call. But it looks like opset10 and opset11 are more complicated--the embedding_bag operation appears to be reimplemented with multiple `g.op` calls and I don't understand those yet. I'll try to figure it out

kurtamohler replied:
After a little bit of investigation, I still don't really understand what has to be done here or how/if tests are being run for opset10 and opset11. Is there documentation that explains any of this? cc @mruberry

mruberry replied:
Unfortunately I don't think there is. cc @bzinodev, @spandantiwari, who would know better

mruberry replied:
Note: One common approach in these situations is to update the non-ONNX code and then file an issue explaining the discrepancy with ONNX.

peterbell10 commented on torch/nn/functional.py:
```diff
@@ -1929,6 +1943,15 @@ def embedding_bag(input, weight, offsets=None, max_norm=None, norm_type=2,
+            assert padding_idx >= -weight.size(0), 'Padding_idx must be within num_embeddings'
+            padding_idx = weight.size(0) + padding_idx
+    else:
+        padding_idx = -1
```
Is there a good reason why `embedding` does this checking in the python level? It seems like it should be simple enough to do in C++ and avoid duplicating the checks in the `torch::` api version.

kurtamohler replied:
I agree that it would be best for the `nn.functional` implementation to defer this check to the C++ level, so I'll make that change.

However, I think the `EmbeddingBag` and `Embedding` modules should still have this check because they call `_fill_padding_idx_with_zero()` upon initialization at the python level. That method simply does this:
```
    def _fill_padding_idx_with_zero(self) -> None:
        if self.padding_idx is not None:
            with torch.no_grad():
                self.weight[self.padding_idx].fill_(0)
```
So it would produce an error by itself if `self.padding_idx` it out of bounds. But I don't think the error produced by the indexing operation would be very clear from the user's perspective.

peterbell10 replied:
Yeah, the module checks are fine. I was meaning the functional interface.

kurtamohler commented:
Thanks @peterbell10 ! Great suggestions as always

kurtamohler commented:
I realized that I did not take `padding_idx` into account in `_embedding_bag_per_sample_weights_backward`, so I need to fix that and add a test for it

mruberry commented on aten/src/ATen/native/EmbeddingBag.cpp:
```diff
@@ -41,12 +41,14 @@ static void make_offset2bag(const Tensor &offsets, Tensor& offset2bag) {
 
-bool isFastPathIndexSelect(const Tensor& src, Tensor& output) {
-  return src.scalar_type() == kFloat && src.stride(1) == 1 && output.stride(1) == 1;
+template<typename index_t>
```
I realize this is an existing issue, but a one-line comment explaining this function and its intended use would be helpful

mruberry commented on aten/src/ATen/native/EmbeddingBag.cpp:
```diff
@@ -41,12 +41,14 @@ static void make_offset2bag(const Tensor &offsets, Tensor& offset2bag) {
 
-bool isFastPathIndexSelectScale(const Tensor& src, const Tensor& scale, Tensor& output) {
-  return src.scalar_type() == kFloat && src.stride(1) == 1 && output.stride(1) == 1 && scale.stride(0) == 1;
+template<typename index_t>
```
One-line comment for this function, too

mruberry commented on aten/src/ATen/native/EmbeddingBag.cpp:
```diff
@@ -41,12 +41,14 @@ static void make_offset2bag(const Tensor &offsets, Tensor& offset2bag) {
-bool isFastPathIndexSelectScale(const Tensor& src, const Tensor& scale, Tensor& output) {
-  return src.scalar_type() == kFloat && src.stride(1) == 1 && output.stride(1) == 1 && scale.stride(0) == 1;
+template<typename index_t>
+bool isFastPathIndexSelectScale(const Tensor& src, const Tensor& scale, Tensor& output, index_t padding_idx) {
```
One-line comment for this function, too

mruberry commented on aten/src/ATen/native/EmbeddingBag.cpp:
```diff
@@ -73,9 +76,11 @@ index_select_add(const Tensor &select_indices,
-    THBlas_axpy<data_t>(ddim, 1,
-            src_data + src_stride0 * select_indices_data[i], src_stride1,
-            output_data + output_stride0 * add_indices_data[i], output_stride1);
+    if (select_indices_data[i] != padding_idx) {
```
For those of us who are not familiar with embedding bag or what the padding_idx argument does (it's documented very cryptically), what's going on here?

A brief comment about why padding_idx is skipped would be helpful here, too

kurtamohler replied:
Does the comment I added make sense?

Perhaps we can make the documentation less cryptic

mruberry commented on aten/src/ATen/native/EmbeddingBag.cpp:
```diff
@@ -159,13 +165,15 @@ index_select_add(const Tensor &select_indices,
-          src_stride1,
-          output_data + output_stride0 * add_indices_data[i],
-          output_stride1);
+      if (select_indices_data[i] != padding_idx) {
```
Add a one-line comment about this condition

mruberry commented on aten/src/ATen/native/EmbeddingBag.cpp:
```diff
@@ -293,11 +305,13 @@ index_select_scale_add(const Tensor &select_indices,
-      auto scale = scale_data[i * scale_stride];
-      for (int64_t j = 0; j < ddim; j++) {
-        output_base[j * output_stride1] += src_base[j * src_stride1] * scale;
+      if (select_indices_data[i] != padding_idx) {
```
One-line comment for this, too

kurtamohler commented:
My recent update changes `padding_idx`'s behavior such that reductions will now ignore the padded words.

I would like to also update the documentation to explain some of the concepts associated with EmbeddingBag in more detail. For instance, I want to explain what a "bag" actually is and why it might be useful.  Perhaps the documentation could also have a concrete example that shows how one could turn a list of sentence strings into a bag of words representation using EmbeddingBag. I think this documentation update should probably be a separate PR though.

kurtamohler commented on aten/src/ATen/native/EmbeddingBag.cpp:
```diff
@@ -59,12 +65,18 @@ index_select_add(const Tensor &select_indices,
-                             bool /*include_last_offset*/) {
+                             bool /*include_last_offset*/,
+                             index_t padding_idx,
+                             Tensor &bag_size) {
```
I'm passing `bag_size` as a non-const reference (here and a few other places) because any time we encounter a padding index, we need to decrement the size for that bag. This might be bad design though. We could instead start with `bag_size` filled with zeros (in all the functions similar to `index_select_add`), and then increment when appropriate.

kurtamohler commented on aten/src/ATen/native/EmbeddingBag.cpp:
```diff
@@ -638,25 +724,30 @@ Tensor _embedding_bag_backward(const Tensor &grad, const Tensor &indices,
   auto nonempty_grad = grad.index_select(0, bag_size.nonzero().view(-1));
 
+  if (padding_idx >= 0) {
+    nonempty_grad.masked_fill_(nonempty_max_indices.eq(padding_idx), 0);
```
I think creating a mask tensor with `.eq()` is probably not ideal. Maybe I can find a more performant way to do this.

jbschlosser replied:
One option is to use `TensorIterator` to write directly to `nonempty_grad`, avoiding the intermediate mask tensor. Something like:
```python
auto iter = TensorIteratorConfig()
  .add_output(nonempty_grad)
  .add_input(nonempty_max_indices)
  .check_all_same_dtype(false)
  .build();
```

Then either use `iter.for_each()` or dispatched CPU / CUDA kernels with a lambda that captures `padding_idx` and checks it against the index.

kurtamohler replied:
Welp, it turns out that the best solution is actually to just remove this change. `max_indices` is never going to contain a padding index--if it did, then the forward mode wouldn't be doing it's job properly. Don't know how I overlooked that lol

kurtamohler commented:
This PR is ready for review. I might need a second opinion on my two most recent comments above

jbschlosser commented on aten/src/ATen/native/cuda/EmbeddingBag.cu:
```diff
@@ -62,43 +63,44 @@ __global__ void EmbeddingBag_updateOutputKernel(
-                per_sample_weights[emb * per_sample_weights_stride]);
-            weightFeatSum += scaleWeightBy * static_cast<accscalar_t>(weightValue);
+        scalar_t weightValue = pad ? static_cast<scalar_t>(0) : weightFeat[weightRow];
+        if (!pad) {
```
I wonder if performance will be affected by warp divergence?

kurtamohler replied:
Yeah you're probably right. Maybe it would be best to remove this `if`, and use ternary operators in the body to make it an effective no-op if `pad == false`, kind of like this:

```
if (mode == MODE_MAX) {
  if (bag_size_ == 0 || weightValue > weightFeatMax) {
    weightFeatMax = pad ? weightFeatMax : weightValue;
    maxWord = pad ? maxWord : input[emb];
  }
} else {
  if (per_sample_weights) {
    accscalar_t scaleWeightBy = static_cast<accscalar_t>(
        per_sample_weights[emb * per_sample_weights_stride]);
    weightFeatSum += scaleWeightBy * static_cast<accscalar_t>(pad ? 0 : weightValue);
  } else {
    weightFeatSum += static_cast<accscalar_t>(pad ? 0 :weightValue);
  }
}
bag_size_ += pad ? 0 : 1;
```

kurtamohler replied:
I wrote a benchmark script: https://github.com/kurtamohler/pytorch-perf-test-scripts/blob/master/embedding_bag-padding_idx/embedding_bag-perf.py

You were right! Performance was down by roughly 30% in the worst case for CUDA, even with `padding_idx=None`.

With the above change, it improves to a drop of about 18%. I'll look for more improvements to see if I can bring it closer to 0%.

kurtamohler replied:
I've improved performance for CUDA and CPU, back up to what it used to be before this PR. Look below for a summary of performance

jbschlosser commented on test/test_nn.py:
```diff
@@ -3353,12 +3353,27 @@ def test_embedding_from_pretrained(self):
+    def test_embedding_bag_from_pretrained(self):
+        a = torch.Tensor([[1, 2, 3], [4, 5, 6]])
+        embedding = nn.EmbeddingBag.from_pretrained(a)
+        self.assertEqual(a, embedding.weight.data)
```
`.data` is old syntax and no longer needed. Please update throughout the newly added tests.

jbschlosser commented on test/test_nn.py:
```diff
@@ -3396,6 +3411,45 @@ def test_embedding_functional(self):
+        res_F = F.embedding_bag(a, embeddings)
+        self.assertEqual(res_old, res_F)
+
+        embed_old = torch.nn.Embedding(4, 3)
```
Should this be `EmbeddingBag`?

jbschlosser commented on test/test_nn.py:
```diff
@@ -3396,6 +3411,45 @@ def test_embedding_functional(self):
+        embed_old = torch.nn.Embedding(4, 3)
+        embed_old = embed_old.from_pretrained(embeddings, padding_idx=2)
+        res_old = embed_old(a)
+        res_F = F.embedding(a, embeddings, padding_idx=2)
```
Should this be `F.embedding_bag`?

jbschlosser commented on test/test_nn.py:
```diff
@@ -3396,6 +3411,45 @@ def test_embedding_functional(self):
+        err_msg = r'padding_idx must be within the number of embeddings'
+        num_embeddings = embeddings.size(0)
+
+        for padding_idx in range(-(num_embeddings + 2), (num_embeddings + 2)):
```
Nice, you knocked out a bunch of the tests I was looking for all at once with this logic :)

jbschlosser commented on test/test_nn.py:
```diff
@@ -11641,6 +11695,178 @@ def test_embedding_padding_idx(self, device, dtype):
+
+            reduction_dim = indices.dim() - 1
+
+            if mode == 'sum' or mode == 'mean':
```
Might be good to add a comment to this block explaining that the `padding_idx` elements are being purposefully removed before the reduction to make the comparison succeed.

jbschlosser commented on test/test_nn.py:
```diff
@@ -11641,6 +11695,178 @@ def test_embedding_padding_idx(self, device, dtype):
+        indices_dim1 = 12
+
+        for mode in ['max', 'mean', 'sum']:
+            for sparse in [False]:
```
Should both sparse and dense be checked?

jbschlosser commented on test/test_nn.py:
```diff
@@ -11641,6 +11695,178 @@ def test_embedding_padding_idx(self, device, dtype):
+                    grad = weights.grad
+
+                    self.assertEqual(grad, grad_check, msg=msg)
+
```
What do you think about adding tests for the all-padding input cases for max, sum, and mean? Or do you think the empty bag cases cover this well enough?

kurtamohler replied:
Good idea, I've added allpad cases to both 1D and 2D input tests

jbschlosser commented on test/test_nn.py:
```diff
@@ -11641,6 +11695,178 @@ def test_embedding_padding_idx(self, device, dtype):
+            return indices_2D
+
+        for mode in ['max', 'mean', 'sum']:
+            for sparse in [False]:
```
Should both sparse and dense be checked?

jbschlosser reviewed (changes requested):
Very nice work! Logic looks correct to me, and the extensive test coverage is solid as well :) Added some minor comments throughout and a few additional testing suggestions.

Most important one is adding an `EmbeddingBag` + `padding_idx` entry to `new_module_tests` in `torch/testing/_internal/common_nn.py` to get Jacobian checks and some other automated testing goodness.

kurtamohler commented on torch/csrc/api/src/nn/modules/embedding.cpp:
```diff
@@ -83,6 +83,15 @@ EmbeddingBagImpl::EmbeddingBagImpl(const EmbeddingBagOptions& options_) : option
 
 void EmbeddingBagImpl::reset() {
+  if (options.padding_idx() != c10::nullopt) {
+    if (*options.padding_idx() > 0) {
```
`options->padding_idx()` isn't the same as `*options.padding_idx()`.

`options->padding_idx()` is equivalent to `(*options).padding_idx()`. Since `options` is not a pointer and doesn't have an dereference operator override, it fails to compile.

`*options.padding_idx()`, on the other hand, is equivalent to `*(options.padding_idx())`. So it's dereferencing the `c10::optional<int64_t>` returned by `padding_idx()`, which has a dereference operator override. Another equivalent is `options.padding_idx().value()`, which I think is more readable, so I'll use that instead

kurtamohler replied:
I'll also change `options.padding_idx() != c10::nullopt` to `options.padding_idx().has_value()`--I think that's more readable as well.

jbschlosser replied:
Indeed, it slipped my mind that `.` has a higher precedence than `*`. I agree with `options.padding_idx().value()` being more readable.

kurtamohler commented on torch/nn/modules/sparse.py:
```diff
@@ -241,6 +241,9 @@ class EmbeddingBag(Module):
                                  supported when ``mode="max"``.
         include_last_offset (bool, optional): if ``True``, :attr:`offsets` has one additional element, where the last element
                                       is equivalent to the size of `indices`. This matches the CSR format.
+        padding_idx (int, optional): If given, pads the embedding vector at index :attr:`padding_idx`. When
```
I'm happy to fix up the documentation. Yeah, probably best to address the non-padding_idx related stuff in a different PR. I'll open a new issue for myself

kurtamohler commented on aten/src/ATen/native/EmbeddingBag.cpp:
```diff
@@ -199,11 +237,18 @@ index_select_scale_add(const Tensor &select_indices,
+      }
+    } else if (bag_size.defined()) {
+      // Decrement bag_size to reflect that the index is padded
+      bag_size_data[add_indices_data[i]]--;
```
This branch is only executed if `mode='sum'` and `include_last_offset == True`, so I'll need to add a test that sets those conditions. But I'm having trouble understanding what `include_last_offset` does. The documentation for `EmbeddingBag` says:

```
include_last_offset (bool, optional): if ``True``, :attr:`offsets` has one additional element, where the last element
                              is equivalent to the size of `indices`. This matches the CSR format
```

While the documentation for `embedding_bag` says:

```
include_last_offset (bool, optional): if ``True``, the size of offsets is equal to the number of bags + 1.
                                      The last element is the size of the input, or the ending index
                                      position of the last bag (sequence).
```

I'm having trouble understanding what either of those mean.

So I decided to just try it out to see what it does. I'm using pytorch 1.7.0 for this:

```
>>> torch.manual_seed(0)
<torch._C.Generator object at 0x7fcd9b439410>
>>> a = torch.randn(5, 4)
>>> a
tensor([[-1.1258, -1.1524, -0.2506, -0.4339],
        [ 0.5988, -1.5551, -0.3414,  1.8530],
        [ 0.4681, -0.1577,  1.4437,  0.2660],
        [ 1.3894,  1.5863,  0.9463, -0.8437],
        [ 0.9318,  1.2590,  2.0050,  0.0537]])
>>> torch.nn.functional.embedding_bag(a, torch.tensor([[0, 1], [3, 4], [2, 3]], dtype=torch.long), mode='sum')
tensor([[-0.5270, -2.7075, -0.5919,  1.4191],
        [ 2.3212,  2.8453,  2.9513, -0.7899],
        [ 1.8575,  1.4286,  2.3900, -0.5776]])
>>> torch.nn.functional.embedding_bag(a, torch.tensor([[0, 1], [3, 4], [2, 3]], dtype=torch.long), mode='sum', include_last_offset=True)
tensor([[-0.5270, -2.7075, -0.5919,  1.4191],
        [ 2.3212,  2.8453,  2.9513, -0.7899]])
```

With the above example, it looks like `include_last_offset=True` just removes the last row from the result, and no `offsets` input was required. Trying it with CUDA, we get something different:

```
>>> torch.nn.functional.embedding_bag(a.cuda(), torch.tensor([[0, 1], [3, 4], [2, 3]], dtype=torch.long).cuda(), mode='sum')
tensor([[-0.5270, -2.7075, -0.5919,  1.4191],
        [ 2.3212,  2.8453,  2.9513, -0.7899],
        [ 1.8575,  1.4286,  2.3900, -0.5776]], device='cuda:0')
>>> torch.nn.functional.embedding_bag(a.cuda(), torch.tensor([[0, 1], [3, 4], [2, 3]], dtype=torch.long).cuda(), mode='sum', include_last_offset=True)
tensor([[-0.5270, -2.7075, -0.5919,  1.4191],
        [ 4.1787,  4.2740,  5.3412, -1.3676]], device='cuda:0')
```

This time, `include_last_offset=True` seems to be combining the last two rows of the output.

But I figured, maybe `include_last_offset` is only meant to work properly if `offsets` is given. But I get the exact same behavior as before:

```
>>> torch.nn.functional.embedding_bag(a, torch.tensor([0, 1, 3, 4, 2, 3], dtype=torch.long), torch.tensor([0, 2, 4]), mode='sum')
tensor([[-0.5270, -2.7075, -0.5919,  1.4191],
        [ 2.3212,  2.8453,  2.9513, -0.7899],
        [ 1.8575,  1.4286,  2.3900, -0.5776]])
>>> torch.nn.functional.embedding_bag(a, torch.tensor([0, 1, 3, 4, 2, 3], dtype=torch.long), torch.tensor([0, 2, 4]), mode='sum', include_last_offset=True)
tensor([[-0.5270, -2.7075, -0.5919,  1.4191],
        [ 2.3212,  2.8453,  2.9513, -0.7899]])
>>> torch.nn.functional.embedding_bag(a.cuda(), torch.tensor([0, 1, 3, 4, 2, 3], dtype=torch.long).cuda(), torch.tensor([0, 2, 4]).cuda(), mode='sum')
tensor([[-0.5270, -2.7075, -0.5919,  1.4191],
        [ 2.3212,  2.8453,  2.9513, -0.7899],
        [ 1.8575,  1.4286,  2.3900, -0.5776]], device='cuda:0')
>>> torch.nn.functional.embedding_bag(a.cuda(), torch.tensor([0, 1, 3, 4, 2, 3], dtype=torch.long).cuda(), torch.tensor([0, 2, 4]).cuda(), mode='sum', include_last_offset=True)
tensor([[-0.5270, -2.7075, -0.5919,  1.4191],
        [ 4.1787,  4.2740,  5.3412, -1.3676]], device='cuda:0')
```

Then it occurred to me that the size of the output does follow the documentation when `include_last_offset=True`: `the size of the offsets is equal to the number of bags + 1`. Since I'm using a size `(3, )` offsets input, we get two bags in the .

But the other part of the description(s) for `include_last_offset` is still confusing to me: `The last element is the size of the input, or the ending index position of the last bag` or `the last element is equivalent to the size of indices`. My interpretation of this is: "`offsets[-1]` indicates the number of elements in the `indices` input". So it seems like there should only be one valid value, `offsets[-1] == indices.numel()`. If so, then I should add an extra element, 6, to the end of the `offsets` inputs from the code samples above. When I do that, I get matching results with `include_last_offset=False` and `include_last_offset=True`.

For CPU:
```
>>> torch.nn.functional.embedding_bag(a, torch.tensor([0, 1, 3, 4, 2, 3], dtype=torch.long), torch.tensor([0, 2, 4]), mode='sum')
tensor([[-0.5270, -2.7075, -0.5919,  1.4191],
        [ 2.3212,  2.8453,  2.9513, -0.7899],
        [ 1.8575,  1.4286,  2.3900, -0.5776]])
>>> torch.nn.functional.embedding_bag(a, torch.tensor([0, 1, 3, 4, 2, 3], dtype=torch.long), torch.tensor([0, 2, 4, 6]), mode='sum', include_last_offset=True)
tensor([[-0.5270, -2.7075, -0.5919,  1.4191],
        [ 2.3212,  2.8453,  2.9513, -0.7899],
        [ 1.8575,  1.4286,  2.3900, -0.5776]])
```

And CUDA gives the same results:
```
>>> torch.nn.functional.embedding_bag(a.cuda(), torch.tensor([0, 1, 3, 4, 2, 3], dtype=torch.long).cuda(), torch.tensor([0, 2, 4]).cuda(), mode='sum')
tensor([[-0.5270, -2.7075, -0.5919,  1.4191],
        [ 2.3212,  2.8453,  2.9513, -0.7899],
        [ 1.8575,  1.4286,  2.3900, -0.5776]], device='cuda:0')
>>> torch.nn.functional.embedding_bag(a.cuda(), torch.tensor([0, 1, 3, 4, 2, 3], dtype=torch.long).cuda(), torch.tensor([0, 2, 4, 6]).cuda(), mode='sum', include_last_offset=True)
tensor([[-0.5270, -2.7075, -0.5919,  1.4191],
        [ 2.3212,  2.8453,  2.9513, -0.7899],
        [ 1.8575,  1.4286,  2.3900, -0.5776]], device='cuda:0')
```

If my interpretation is correct, I still don't understand a few important things:

* Why am I allowed to set `include_last_offset=True` if I don't give an `offsets` input?
* Is the behavior consistent with the confusing name `include_last_offset`?
* Why am I allowed to set `offsets[-1] != indices.numel()`?
* What purpose does this feature serve?

One of the two descriptions of `include_last_offset` says `This matches the CSR format`. Still, I don't see what `include_last_offset` offers if the inputs are sparse. But perhaps it would become clear if I had a better grasp on sparse tensors.

What's your take on this @jbschlosser ?

kurtamohler replied:
Sorry for the long post, I'm not sure if I could describe my confusion more concisely :stuck_out_tongue_closed_eyes:

kurtamohler replied:
Ah ok, I found this issue explaining why `include_last_offset` exists: #29019

Its whole purpose is to avoid needing a special case for the last element of `offsets` when `embedding_bag` calculates the size of each bag, when `input` is 1-D. This apparently speeds up parallelism. It seems like `offsets[-1]` should always be equal to `input.size(0)` if `include_last_offset=True`, so I filed #52851 to fix that.

Anyway, I now understand what `include_last_offset` is supposed to do, so I'll add test cases to exercise these branches.

kurtamohler replied:
Yes, `include_last_offset` is only used in the 1D case, so later on we could also consider adding an error if it's given for the 2D case

kurtamohler commented on torch/csrc/api/include/torch/nn/options/embedding.h:
```diff
@@ -115,6 +115,10 @@ struct TORCH_API EmbeddingBagOptions {
+  /// If given, pads the embedding vector at index `padding_idx`. When
+  /// a `padding_index` is encountered in `input` during a reduction,
+  /// it is skipped. This allows each bag to be a different logical size.
+  TORCH_ARG(c10::optional<int64_t>, padding_idx) = c10::nullopt;
```
`torch/csrc/api/src/nn/options/embedding.cpp` just has the constructor definition for `EmbeddingBagOptions`, and optional arguments like `padding_idx` are not part of the constructor.

kurtamohler commented on test/test_nn.py:
```diff
@@ -3353,12 +3353,27 @@ def test_embedding_from_pretrained(self):
         output = embedding(input)
         self.assertEqual(a, output)
 
+    def test_embedding_bag_from_pretrained(self):
```
Done

kurtamohler commented on torch/nn/modules/sparse.py:
```diff
@@ -284,9 +287,18 @@ class EmbeddingBag(Module):
         tensor([[-0.8861, -5.4350, -0.0523],
                 [ 1.1306, -2.5798, -1.0044]])
+
+        >>> # example with padding_idx
```
Good idea, I've added it

kurtamohler commented:
I wrote a performance measurement script. The `*.sh` files in this directory run the script with CUDA, CPU single core, and CPU 4 core. I used taskset for CPU to pin to specific cores, to minimize context switching: https://github.com/kurtamohler/pytorch-perf-test-scripts/tree/master/embedding_bag-padding_idx

Currently, all the `padding_idx=None` cases have pretty much equal performance compared to master. Here is a spreadsheet with runtime measurements, including the speedup of this PR over master: https://docs.google.com/spreadsheets/d/1UvM-UaECXDrle0U4K7pzM57r895BKmWgmhg0dA8kskg/edit?usp=sharing

For clarity, here is an explanation of a couple of the columns in the sheet:
* include_backward - If "TRUE", time measurement includes both forward and backward
* padding_ratio - This is the ratio of padding in the input to `embedding_bag`. If zero, nothing is padded and `padding_idx=None`. Otherwise, `padding_idx` is set to 0, and the appropriate number of elements in the input is set to 0 randomly to match the padding ratio. For instance, if padding_ratio is 1, everything is set to 0, so it's all padded

There is a noteworthy case, where the reduction mode is either `sum` or `mean` for CPU float32. `embedding_bag` CPU has a "fastpath" implementation for this case, but only when `padding_idx=None`. Otherwise, the fastpath cannot be used, so the runtime jumps significantly higher with `padding_idx!=None`, compared to `padding_idx=None`. So its best to look at the corresponding float64 cases, which do not have a fastpath. We can see that the CPU `sum/mean` float64 case is comparable to master, for all values of `padding_idx`. Therefore, the jump in runtime for CPU float32 sum/mean when `padding_idx!=None` is very likely just because it isn't using fastpath, not because of any performance degradation that I added.

I should point out that one of the reasons I was able to match the original performance for CUDA was because I split the `EmbeddingBag_updateOutputKernel` function (which originally handled max, sum, and mean) in `aten/src/ATen/native/cuda/EmbeddingBag.cu` into `EmbeddingBag_updateOutputKernel_max` and `EmbeddingBag_updateOutputKernel_sum_mean`. Separating these implementations avoids many repeated conditional statements checking the reduction mode within the function's loop, saving a fair amount of time.

kurtamohler commented:
@jbschlosser, this is ready for a re-review when you get a chance

jbschlosser commented:
> @jbschlosser, this is ready for a re-review when you get a chance

Awesome, will try to get to it a little later this week.

neginraoof commented:
Hi @kurtamohler,
Thanks for this update. I think as you are updating this torch operator, we need to also update the ONNX conversion behavior for embedding_bags. Could you please tell us the plan for landing this PR so that we plan the exporter update accordingly?
Thanks.

kurtamohler commented:
Hi @neginraoof, as far as I know, the plan is to land this PR whenever it's approved. BTW I updated the `torch/onnx/symbolic_opset*.py` files to accept the `padding_idx` argument, and I made them just throw an error. I attempted to enable the argument fully, but I couldn't figure out how.

jbschlosser commented on test/test_nn.py:
```diff
@@ -12041,6 +12100,215 @@ def test_embedding_padding_idx(self, device, dtype):
+    # padding indices to fill in the gaps indicated by the offset array
+
+    @onlyOnCPUAndCUDA
+    @dtypes(torch.float32, torch.float64)
```
I think you can also use `dtypesIfCUDA` to test the `half and `bfloat16` cases

jbschlosser reviewed (commented):
Finally got around to reviewing the most recent changes! It's looking really good- I especially appreciate the extensive test coverage. Couple of minor things but I think we're getting close to wrapping this up :) Thanks for all the hard work on this!

kurtamohler commented on test/test_nn.py:
```diff
@@ -12041,6 +12100,215 @@ def test_embedding_padding_idx(self, device, dtype):
+                continue
+
+            if allpad:
+                torch.empty(indices_dim0, indices_dim1, dtype=torch.long, device=device).fill_(1)
```
Ah darn, thanks for catching that

kurtamohler commented:
Ready for another look when you have a minute, @jbschlosser

jbschlosser reviewed (approved):
Hm, something doesn't seem quite right:

```python
import torch

emb = torch.nn.EmbeddingBag.from_pretrained(torch.tensor([[1., 1.], [2., 2.], [3., 3.]]), padding_idx=1, mode='mean')
print('weight:', emb.weight)
input = torch.tensor([[0, 1, 2]])
print('input:', input, input.shape)
print('output:', emb(input))
```
```
weight: Parameter containing:
tensor([[1., 1.],
        [0., 0.],
        [3., 3.]])
input: tensor([[0, 1, 2]]) torch.Size([1, 3])
output: tensor([[1.3333, 1.3333]])
```

I'm passing 2D indices of shape (1, 3) (one bag of length 3 with one padding element) and the output does not seem to be divided by the correct number of non-padding elements (2). Am I missing something?

Also, I guess we need a separate issue for a future include_padding=False arg to control whether padding embedding vectors are part of the reduction?

kurtamohler commented:
Well that's interesting, thanks for catching that @jbschlosser. It seems like it works as expected if `input.size(0) > 1`, but breaks if it's 1:

```
>>> torch.nn.functional.embedding_bag(torch.tensor([[0, 1, 2]]), torch.tensor([[1., 1.], [2., 2.], [3., 3.]]), padding_idx=1, mode='mean')
tensor([[1.3333, 1.3333]])
>>> torch.nn.functional.embedding_bag(torch.tensor([[0, 1, 2], [0, 1, 2]]), torch.tensor([[1., 1.], [2., 2.], [3., 3.]]), padding_idx=1, mode='mean')
tensor([[2., 2.],
        [2., 2.]])
```

Which is weird--I don't remember seeing anything in the code that makes a special case for this. In any case, I'll fix it and add a test case that exercises it, since all of my tests currently set `input.size(0) > 1`.

kurtamohler commented:
Alright, @jbschlosser, I fixed that issue and added test cases where dim0 of the input is 1

jbschlosser reviewed (commented):
Confirmed fix for indices of shape (1, *)! Sorry to do this to you, but there's one last thing that just came up that should be addressed before we finish this out.

jbschlosser reviewed (commented):
lgtm!

kurtamohler committed:
Remove unnecessary args and variables

kurtamohler committed:
Merge branch 'master' into EmbeddingBag-padding_idx-3194

kurtamohler commented:
@jbschlosser I've removed the unnecessary captures and some function arguments

kurtamohler committed:
Remove accidental comments

jbschlosser commented:
> The merge conflict is from a recent commit that makes significant internal changes to embedding_bag. I'm in the process of merging, and I've gotten it to a point where it builds. But a lot of the tests are failing, so it might take me a little while to fix

@kurtamohler Sounds good! Let me know when that's finished up, and there may be some additional work on the internal side I'll need to figure out.

kurtamohler committed:
Merge branch 'master' into EmbeddingBag-padding_idx-3194

kurtamohler committed:
Merge branch 'master' into EmbeddingBag-padding_idx-3194

kurtamohler committed:
Add padding_idx to new JIT functor

kurtamohler committed:
Merge branch 'master' into EmbeddingBag-padding_idx-3194

kurtamohler commented:
@jbschlosser, finished fixing the conflicts!

jbschlosser commented:
To get the BC check tests passing, this line can be added to the `allow_list` in `test/backward_compatibility/check_backward_compatibility.py`:
```python
("aten::_embedding_bag_dense_backward", datetime.date(9999, 1, 1)),
```

kurtamohler committed:
Allow BC-break for _embedding_bag_dense_backward

kurtamohler committed:
Merge branch 'master' into EmbeddingBag-padding_idx-3194

kurtamohler commented:
> this line can be added to the `allow_list`

Done

jbschlosser commented:
Hey @kurtamohler, is it cool if we move the quantization stuff to a different PR? There's a bunch of internal builds / tests failing because of the quant changes and I'd like to get this merged in and take care of that later on.

kurtamohler commented:
@jbschlosser, sure thing, I'll undo those changes

kurtamohler committed:
Undo changes to quantized embedding_bag

kurtamohler committed:
Merge branch 'master' into EmbeddingBag-padding_idx-3194

kurtamohler committed:
Revert additional quantization changes

kurtamohler commented:
CI failures seem to be coming from the fact that the quantized embedding_bag does not have the padding_idx arg anymore. I'm not sure yet how to fix that, will look into it tonight or tomorrow

kurtamohler committed:
Throw error if quantized embedding_bag's padding_idx is not None

kurtamohler committed:
Merge branch 'master' into EmbeddingBag-padding_idx-3194

kurtamohler commented on torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:
```diff
@@ -375,7 +375,10 @@ Node* insertEmbeddingBagOps(Node* observer, const std::string& op_name) {
-  qembedding_bag_inputs.push_back(embedding_bag_inputs[inputs_size - 1]);
+  qembedding_bag_inputs.push_back(embedding_bag_inputs[inputs_size - 2]);
+
+  TORCH_CHECK(embedding_bag_inputs[inputs_size - 1]->mustBeNone(),
```
@jbschlosser, as far as I could figure out so far, these are the minimum amount of changes I had to make with respect to quantization for this PR. All this does is throw an error if quantization is attempted on an embedding bag with a padding_idx defined. Note, I also added a test in `test/quantization/test_quantize_jit.py` to make sure the error is raised correctly

jbschlosser replied:
Awesome, thanks for taking the time to get this straightened out! Will attempt the merge again now.

kurtamohler committed:
Format fixes

kurtamohler committed:
Additional format fix

kurtamohler committed:
Fix incorrect index type

kurtamohler committed:
Merge branch 'master' into EmbeddingBag-padding_idx-3194

kurtamohler committed:
Merge branch 'master' into EmbeddingBag-padding_idx-3194

kurtamohler commented:
@jbschlosser, some new merge conflicts appeared, and I've fixed them

neginraoof commented:
@kurtamohler Quick question about this argument. I see recently the purpose of padding_idx has changed for Embedding op, and it is no longer used to pad the output embedding vectors. But it is used to skip weight updates in training for certain indices.
Is this change inline with the recent updated? Is padding_idx used in the forward pass here?

kurtamohler commented:
@neginraoof, thanks for letting me know. In this PR, `padding_idx` is used in the forward pass. However, it doesn't actually set the output values to 0. Instead, the reduction performed within `embedding_bag` ignores the values at `padding_idx`. The grad for padded elements will be 0, so they will be skipped in training.

To me, it seems like this PR is still conceptually and functionally compatible with the updated Embedding (@jbschlosser, do you agree?).

Some of the tests in this PR check to make sure that `embedding_bag` with `padding_idx` defined gives the same result as `embedding` followed by a reduction that ignores the padded values. In the forward pass for these, it shouldn't actually matter whether or not `embedding` sets the values to 0, since the reduction is ignoring those values. But I will rebase and rerun tests to make sure.

jbschlosser commented:
> @kurtamohler Quick question about this argument. I see recently the purpose of padding_idx has changed for Embedding op, and it is no longer used to pad the output embedding vectors. But it is used to skip weight updates in training for certain indices.
> Is this change inline with the recent updated? Is padding_idx used in the forward pass here?

@neginraoof Quick clarification: the `padding_idx` arg for `Embedding` was always intended as a way to keep a certain embedding vector fixed (i.e. keep it from receiving gradient updates). In practice, this value is often fixed to all zeros, but certain use cases require non-zero embedding vectors for padding. The idea of "padding the output embedding vectors" is not really an accurate description of this behavior, although the docs changed to read that later on.

In fact, due to this new description, there was a regression in 1.8.0 where the purpose of the arg was misinterpreted and its behavior was changed to *set* the vector at `padding_idx` to all zeros in certain conditions (see https://github.com/pytorch/pytorch/issues/46585). This caused some mayhem among people who need non-zero, fixed padding embedding vectors (see https://github.com/pytorch/pytorch/issues/53368), and the regression was fixed (see https://github.com/pytorch/pytorch/pull/53447).

> To me, it seems like this PR is still conceptually and functionally compatible with the new Embedding behavior (@jbschlosser, do you agree?).

Since `padding_idx` for `EmbeddingBag` indicates that those items should be left out from the reduction (thus not receiving gradient updates), I think this is conceptually and functionally compatible with the correct `Embedding` + `padding_idx` behavior. In both cases, `padding_idx` is used to keep embedding vectors from receiving gradient updates (and perhaps this should be made more clear in the `EmbeddingBag` docs).

Note that this PR currently only serves those who need all-zero padding embedding vectors, as it is generally *not* desired for the zeros to take part in the reduction (since that would skew the resultant values). @kurtamohler and I have discussed a future (default False) flag that would *include* padding embedding vector elements in the reduction, while still keeping the vector from receiving gradient updates. This would serve the use cases of those who need fixed, non-zero padding embedding vectors. I believe this use case is less common than the all-zero padding vector case, but given the backlash from the `Embedding` regression, certainly still important. Once this flag is in, enabling it will result in wholly equivalent behavior as `Embedding` + `padding_idx` + a reduction.

PS @kurtamohler I haven't forgotten about this PR - I'm actively working on an update for some internal code that unfortunately depends on the pre-`padding_idx` signature of `EmbeddingBag` too heavily. Once that's in, we'll be able to merge this.

kurtamohler commented:
Great, thanks @jbschlosser !

kurtamohler committed:
Merge branch 'master' into EmbeddingBag-padding_idx-3194

neginraoof commented:
@kurtamohler @jbschlosser Thanks a lot for clarification. So, what I conclude is that in both cases this flag is used to avoid gradient updates for embedding vectors. For Embedding, it's used in backward pass only, but in EmbeddingBag it is used in forward pass to exclude items from reduction.
Thanks for updating ONNX accordingly to error out when padding_idx is set for EmbeddingBag.

kurtamohler committed:
Merge branch 'master' into EmbeddingBag-padding_idx-3194

kurtamohler committed:
Merge branch 'master' into EmbeddingBag-padding_idx-3194

kurtamohler committed:
Merge branch 'master' into EmbeddingBag-padding_idx-3194

jbschlosser commented:
@kurtamohler Update: sorry for the long delay! I've found a workaround to the internal BC problems. It involves having two `embedding_bag` entries in `native_functions.yaml`, with one matching the old signature and one matching the signature with the new `padding_idx` arg. The latter one will require all args to be specified with no defaults to avoid ambiguous calls.

Happy to clarify further in slack, or alternatively I can take care of it by pushing directly to this branch if you're cool with that!

kurtamohler committed:
Restore original signature for compatibility

kurtamohler committed:
Merge branch 'master' into EmbeddingBag-padding_idx-3194

kurtamohler commented:
@jbschlosser, I've added the original signature back in. Hopefully didn't break anything in CI, we'll see!

kurtamohler committed:
Merge branch 'master' into EmbeddingBag-padding_idx-3194

albanD commented on tools/autograd/derivatives.yaml:
```diff
@@ -1280,13 +1280,13 @@
+  per_sample_weights: _embedding_bag_per_sample_weights_backward(grad, weight, indices, offsets, result1, mode, padding_idx)
 
-- name: _embedding_bag_dense_backward(Tensor grad, Tensor indices, Tensor offsets, Tensor offset2bag, Tensor bag_size, Tensor maximum_indices, int num_weights, bool scale_grad_by_freq, int mode, Tensor? per_sample_weights) -> Tensor
+- name: _embedding_bag_dense_backward(Tensor grad, Tensor indices, Tensor offset2bag, Tensor bag_size, Tensor maximum_indices, int num_weights, bool scale_grad_by_freq, int mode, Tensor? per_sample_weights, int padding_idx=-1) -> Tensor
```
FYI that you didn't remove the derivative formula corresponding to offsets, so that's malformed now.
I'm fixing it.

kurtamohler replied:
Darn, thanks for catching that!
